Startup fails with Undertow when binding to 0.0.0.0 on a dual-stack host

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. clone https://github.com/ianroberts/micronaut-undertow-startup-failure on a machine that has dual-stack IPv4/v6 networking
2. build the app (`./gradlew shadowJar`)
3. run the app with `java -jar`

### Expected Behaviour

App starts successfully

### Actual Behaviour

```
17:36:12.257 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: For input string: "0:0:0:0:0:0:0:8080"
io.micronaut.http.server.exceptions.InternalServerException: For input string: "0:0:0:0:0:0:0:8080"
	at io.micronaut.servlet.undertow.UndertowServer.getURL(UndertowServer.java:108)
	at io.micronaut.runtime.Micronaut.lambda$start$2(Micronaut.java:95)
	at java.util.Optional.ifPresent(Optional.java:159)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:311)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:297)
	at com.example.Application.main(Application.java:8)
Caused by: java.net.MalformedURLException: For input string: "0:0:0:0:0:0:0:8080"
	at java.net.URL.<init>(URL.java:627)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at java.net.URI.toURL(URI.java:1089)
	at io.micronaut.servlet.undertow.UndertowServer.getURL(UndertowServer.java:106)
	... 6 common frames omitted
Caused by: java.lang.NumberFormatException: For input string: "0:0:0:0:0:0:0:8080"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at java.net.URLStreamHandler.parseURL(URLStreamHandler.java:222)
	at java.net.URL.<init>(URL.java:622)
	... 10 common frames omitted
```

### Environment Information

- **Operating System**: Mac OS 10.15, but same has been observed on Ubuntu 20.04 under kubernetes
- **Micronaut Version:** micronaut 2.3.0 or 2.2.1 with `micronaut-http-server-undertow` v2.1.2
- **JDK Version:** OpenJDK 8

### Example Application

- https://github.com/ianroberts/micronaut-undertow-startup-failure

### Workarounds

Any of the following will avoid the error:

- Do not set `micronaut.server.host` in application.yml (though then the app binds only to localhost by default)
- Disable IPv6 with `-Djava.net.preferIPv4Stack=true`
- Configure the logger for `io.micronaut.runtime.Micronaut` to be `ERROR` or higher
  - this fixes it because the failing call to `UndertowServer.getURL` is guarded by an `if(LOG.isInfoEnabled())`

### Analysis

The root cause appears to be that in this case `UndertowServer.getHost()` is returning the string form of an IPv6 all-zeros address, which `getURI` blindly concatenates in `URI.create(getScheme() + "://" + getHost() + ":" + getPort())` without wrapping it in square brackets.  Safer would be to use the multi-argument `java.net.URI` constructor, which handles all necessary escaping:

```
new URI(getScheme(), null, getHost(), getPort(), null, null, null)
```

--- a/http-server-undertow/src/main/java/io/micronaut/servlet/undertow/UndertowServer.java
+++ b/http-server-undertow/src/main/java/io/micronaut/servlet/undertow/UndertowServer.java
@@ -111,7 +111,11 @@ public class UndertowServer extends AbstractServletServer<Undertow> {
 
     @Override
     public URI getURI() {
-        return URI.create(getScheme() + "://" + getHost() + ":" + getPort());
+        try {
+            return new URI(getScheme(), null, getHost(), getPort(), null, null, null);
+        } catch(URISyntaxException e) {
+            throw new InternalServerException(e.getMessage(), e);
+        }
     }
 
     @Override

--- a/http-server-undertow/src/main/java/io/micronaut/servlet/undertow/UndertowServer.java
+++ b/http-server-undertow/src/main/java/io/micronaut/servlet/undertow/UndertowServer.java
@@ -113,7 +113,7 @@ public class UndertowServer extends AbstractServletServer<Undertow> {
     public URI getURI() {
         try {
             return new URI(getScheme(), null, getHost(), getPort(), null, null, null);
-        } catch(URISyntaxException e) {
+        } catch (URISyntaxException e) {
             throw new InternalServerException(e.getMessage(), e);
         }
     }

